app/src/main/java/com/example/bottomnav/PrimitiveDecoder.java

```
@@ -1,54 +1,43 @@
 package com.example.bottomnav;
 
-public class PrimitiveDecoder<T> implements DataDecoder {
-    protected VariableContents contents;
-    protected T rawValue;
-    protected String value;
-    protected int size = 1;
-
-    public PrimitiveDecoder(VariableContents can) {
-        contents = can;
-    }
+import java.nio.ByteBuffer;
+import java.util.Optional;
 
-    @Override
-    public String decode(Integer canId, byte[] payload) {
-        return null;
-    }
-
-    @Override
-    public int getSize() {
-        return size;
-    }
-
-    @Override
-    public String valueToString() {
-        return "" + contents.name + ": " + value;
-    }
+public class PrimitiveDecoder<T> implements DataDecoder {
+    protected int packetSize;
+    protected String variableName;
 
-    @Override
-    public T valueToRaw() {
-        return rawValue;
+    public PrimitiveDecoder(int size, String name) {
+        packetSize = size;
+        variableName = name;
     }
 
     @Override
-    public String getVarNameAt(int i) {
-        return contents.name;
-    }
-
-    public String getVarName() {
-        return contents.name;
+    public Optional<T> decodeToRaw(byte[] payload) {
+        if (payload.length < packetSize) {
+            return Optional.empty();
+        }
+        return getRawValue(payload);
     }
 
     @Override
-    public String getValueStringAt(int i) {
-        return value;
+    public Optional<String> decodeToString(byte[] payload) {
+        Optional<T> rawOptional = decodeToRaw(payload);
+        if (rawOptional.isPresent()) {
+            return Optional.of(variableName + ": " + rawOptional.get());
+        } else {
+            return Optional.empty();
+        }
     }
 
-    public String getValueString() {
-        return value;
+    public Optional<T> getRawValue(byte[] payload) {
+        return null;
     }
 
-    public int getPacketSize() {
-        return contents.packetSize;
+    public byte[] wrapPayload(byte[] payload) {
```

Can you add a comment on the behavior of this and what this is for?

I'm also not convinced it's a good idea to trim and reformat data, in part because it may give the user misleading information about data being cleaner and more well formatted than it actually might be on the wire. We already have the infrastructure to fail noisily (but not catastrophically) in the case of bad input using Optional.empty.